README
= User notes =

The software-center application aims to make the software handling on
the computer easy and consistent.

= Developer notes =

All non UI code must come with tests in the test/ subdirectoy.

To setup your development environment, you'll need to ensure the following
extra packages are installed:

sudo apt-get install xvfb python-coverage python-mock python-aptdaemon.test \
     python-qt4 python-unittest2 python-lxml pep8
sudo apt-get build-dep software-center

You can then run tests with:

cd test;make

You can run a developer instance with:

python setup.py build
./software-center

The initial launch of this will take a bit as it will build a private
search database but this is only needed once.

== query parser ==

The query parser understands :
 "pkg:2vcard", "mime:text/html", "section:web", "origin:main"
prefixes.

== aptdaemon ==
 * the dbus limits for the system bus are rather low, this means that
   adding  <limit name="max_match_rules_per_connection">512</limit>
   and using something bigger than 512 is a good idea

== environment ==

The following environment variables are supported:

SOFTWARE_CENTER_AGENT_HOST - an alternative host to query for pay software
SOFTWARE_CENTER_REVIEWS_HOST - an alternative host for the ratings&reviews
SOFTWARE_CENTER_DEBUG_HTTP - enable httplib2 debuging
SOFTWARE_CENTER_DEBUG_WEBKIT - enable webkit debuging
SOFTWARE_CENTER_IPSUM_REVIEWS - generate random reviews
SOFTWARE_CENTER_FAKE_REVIEW_API - use a fake server for all review network operations
SOFTWARE_CENTER_GWIBBER_MOCK_USERS=2 - use mock gwibber service
SOFTWARE_CENTER_AGENT_INCLUDE_QA - show not yet QA apps available from the agent
SOFTWARE_CENTER_NET_DISCONNECTED - make software-center's netstatus module believe network manager is in a disconnected state
SOFTWARE_CENTER_WEBLIVE_HOST - overwrite default weblive server
SOFTWARE_CENTER_DISTRO_CODENAME - overwrite "lsb_release -c -s" output
SOFTWARE_CENTER_ARCHITECTURE - overwrite the current architecture
SOFTWARE_CENTER_NO_SC_AGENT - disable the software-center-agent
SOFTWARE_CENTER_DISABLE_SPAWN_HELPER - disable everything that is run via the "SpawnHelper", i.e. recommender-agent, software-center-agent, reviews
SOFTWARE_CENTER_DEBUG_TABS - show notebook tabs for debugging
SOFTWARE_CENTER_FORCE_DISABLE_CERTS_CHECK - disables certificates checking in webkit views (for use in test environments)
SOFTWARE_CENTER_FORCE_NON_SSL - disable SSL (for use in test environments)

== applications.menu ==

The menu file parser understands:
Category, And, Or, Not

The following additional XML filters are definied:
SCType - e.g. "Applicatin"
SCChannel - e.g. "lucid-partner"
SCSection - e.g. "net"
SCPkgname - e.g. "gimp"

Additional .menu files can be added in:
/usr/share/app-install/menu.d
that software-center will read and parse.

== XAPIAN ==

The following special prefixes are used:

AA - application name (Abiword)
AP - package name (abiword)
AS - archive pocket (main)
AE - archive section (mail, base, ...)
AC - category (AudioVideo)
AM - MimeType (application/x-ogg)
AT - type (Application)
AH - channel


The following values are used:

XAPIAN_VALUE_PKGNAME - pkgname
XAPIAN_VALUE_ICON - icon name
XAPIAN_VALUE_GETTEXT_DOMAIN - gettext domain
XAPIAN_VALUE_ARCHIVE_SECTION - archive section (main, restricted, universe, multiverse)
XAPIAN_VALUE_ARCHIVE_ARCH - architectures (seperated with ",", e.g. i386,amd64) - may be empty
XAPIAN_VALUE_POPCON - popcon data
XAPIAN_VALUE_SUMMARY - summary text
XAPIAN_VALUE_DESKTOP_FILE - the desktop file that the information comes from
XAPIAN_VALUE_PRICE - the price (if its a for-pay app)
XAPIAN_VALUE_ARCHIVE_CHANNEL - channel (third party)
XAPIAN_VALUE_ARCHIVE_PPA - the PPA name that the application is in
XAPIAN_VALUE_ARCHIVE_DEBLINE - a deb line for the sources.list to access the given app
XAPIAN_VALUE_ARCHIVE_SIGNING_KEYID - signing key id for the repository
XAPIAN_VALUE_PURCHASED_DATE - the data a for-pay app was purchased (only available after the software-center-agent server was queried)
XAPIAN_VALUE_SCREENSHOT_URLS - a (optional) list of "," seperated screenshot urls that overrides the default
XAPIAN_VALUE_ICON_NEEDS_DOWNLOAD - icon needs to be fetched
XAPIAN_VALUE_THUMBNAIL_URL - thumbnail url



debian/changelog
software-center (13.10-0ubuntu4.1) trusty-proposed; urgency=low

  * debian/patches/06_lp1308354.patch:
    - disable paste when the search entry is not visible (LP: 1308354)

 -- Michael Vogt <michael.vogt@ubuntu.com>  Thu, 17 Apr 2014 15:43:28 +0200

software-center (13.10-0ubuntu4) trusty; urgency=low

  [ Robert Ancell ]
  * debian/patches/04_bzr_r3322_lp1281317.patch
    - merge upstream rev3322 to fix LP: #1281317
  
  [ Michael Vogt ]
  * debian/patches/05_lp225885.patch:
    - cherry pick  lp:~mvo/software-center/lp1225885 to fix LP: #1225885

 -- Michael Vogt <michael.vogt@ubuntu.com>  Thu, 10 Apr 2014 18:21:54 +0200

software-center (13.10-0ubuntu3) trusty; urgency=medium

  * debian/patches/03_bzr-save-settings.patch:
    - Fix parent object initialization chaining. (LP: #1244493)

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Fri, 21 Mar 2014 14:06:18 -0400

software-center (13.10-0ubuntu2) trusty; urgency=low

  * debian/patches/01_bzr-typeerror.patch:
    - Avoid creating a launcher without .desktop files. (LP: #1227277)
  * debian/patches/02_bzr-drop-distroseries.patch:
    - Drop the use of DISTROSERIES. (LP: #1254892)

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Tue, 26 Nov 2013 10:16:14 -0500

software-center (13.10-0ubuntu1) saucy; urgency=low

  * New upstream release.
  * debian/patches/00_bzr-no-network-local-debs.patch:
    - Don't require network for installing local debs. (LP: #926763)

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Tue, 08 Oct 2013 16:03:35 -0400

software-center (13.09-0ubuntu1) saucy; urgency=low

  * New upstream release.
    - Improved integration with Unity launcher.
    - Use "Ubuntu One" authorization rather than separate token.
    - Fix warnings about old style classes.

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Tue, 17 Sep 2013 14:56:13 -0400

software-center (13.08-0ubuntu1) saucy; urgency=low

  * New upstream release.
  * debian/patches:
    - Remove upstreamed patches.

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Wed, 07 Aug 2013 14:57:08 -0400

software-center (13.07-0ubuntu4) saucy; urgency=low

  * debian/control, debian/tests:
    - Remove the autopkgtests as they have never worked, and are
      very broken. We might add them back later.

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Thu, 18 Jul 2013 09:50:31 -0400

software-center (13.07-0ubuntu3) saucy; urgency=low

  * debian/patches/02_bzr-more-test-fixes.patch:
    - Remove another test that requires network.
    - Fix the pyflakes test to use specific sub-directories.

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Wed, 17 Jul 2013 11:41:25 -0400

software-center (13.07-0ubuntu2) saucy; urgency=low

  * debian/patches/01_bzr-test-fixes.patch:
    - Remove the bad tests files instead of renaming them.

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Tue, 16 Jul 2013 14:11:34 -0400

software-center (13.07-0ubuntu1) saucy; urgency=low

  * New upstream release.
    - Use proper __init__ of parent hierarchy call. (LP: #1198822)
  * debian/patches/01_bzr-fix-tests.patch:
    - Include patch from upstream to fix some more tests.

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Fri, 12 Jul 2013 16:40:52 -0400

software-center (13.05-0ubuntu3) UNRELEASED; urgency=low

  * Test fixes:
    - Add missing test Depends: sudo, python-distutils-extra, apt-xapian-index
    - Set XDG_CACHE_HOME to a writable directory ($ADTTMP)
    - Add restrictions: rw-build-dir since the test runner creates files
    - Disable a few failing tests
    - Appease pyflakes
  * Drop patch in this upstream release. 

 -- Iain Lane <iain.lane@canonical.com>  Tue, 09 Jul 2013 11:54:43 +0100

software-center (13.05-0ubuntu2) saucy; urgency=low

  * debian/patches/01_bzr-fix-1038429.patch:
    - Don't use MI, and use super to call the parent __init__. (LP: #1038429)

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Fri, 05 Jul 2013 17:39:12 -0400

software-center (13.05-0ubuntu1) saucy; urgency=low

  * New upstream release.
  * debian/patches:
    - Remove patch files which are included upstream.
  * debian/watch:
    - Update watch file for new series.

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Wed, 05 Jun 2013 16:26:46 -0400

software-center (5.6.0-0ubuntu4) saucy; urgency=low

  * debian/patches/saucy-known-series.patch: Add new patch to know about
    (Ubuntu) 'saucy', fixing FTBFS. 

 -- Iain Lane <iain@orangesquash.org.uk>  Wed, 01 May 2013 18:52:16 +0100

software-center (5.6.0-0ubuntu3) raring-proposed; urgency=low

  * debian/software-center.postinst:
    - Only run update-software-center if python-aptdaemon is configured,
      to avoid crashing on upgrade, and to work around Debian 671711.
      (LP: #1173249)

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Fri, 26 Apr 2013 14:11:50 -0400

software-center (5.6.0-0ubuntu2) raring; urgency=low

  * debian/patches/00_bzr-typerror-remove-menuitem.patch:
    - Avoid TypeError caused by external bug, by removing the Cut
      menu item from the main menu, and making the Copy item not
      used for the search entry. (LP: #1151621)

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Mon, 08 Apr 2013 15:04:00 -0400

software-center (5.6.0-0ubuntu1) raring; urgency=low

  * New upstream release.

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Wed, 27 Mar 2013 15:51:28 -0400

software-center (5.5.6-0ubuntu2) raring; urgency=low

  * debian/patches/00_bzr-run-tests-setup.patch:
    - Run setup.py build before the tests get run; clean after. (LP: #1158290)

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Fri, 22 Mar 2013 09:48:37 -0400

software-center (5.5.6-0ubuntu1) raring; urgency=low

  * New upstream release.
    - Remove "foo" tooltip from account login/creation dialogs. (LP: #1076189)
    - Add new icons from design team. (LP: #1081691)
    - Remove extra .desktop file that we don't need. (LP: #1153548)
  * debian/rules:
    - Remove complicated unbranding rule.

 -- Rodney Dawes <rodney.dawes@ubuntu.com>  Wed, 20 Mar 2013 15:52:01 -0400


softwarecenter/enums.py
# Copyright (C) 2009 Canonical
#
# Authors:
#  Michael Vogt
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; version 3.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

import os
from gettext import gettext as _

# pkgname of this app itself (used for "self-awareness", see spec)
SOFTWARE_CENTER_PKGNAME = 'software-center'

# name of the app in the keyring, untranslated, see bug #773214 for the
# rational
SOFTWARE_CENTER_NAME_KEYRING = "Ubuntu Software Center"
SOFTWARE_CENTER_SSO_DESCRIPTION = _(
    "To reinstall previous purchases, sign in to the "
    "Ubuntu Single Sign-On account you used to pay for them.")

SOFTWARE_CENTER_DEBUG_TABS = os.environ.get(
    'SOFTWARE_CENTER_DEBUG_TABS', False)

SOFTWARE_CENTER_BUY_HOST = os.environ.get(
    "SOFTWARE_CENTER_BUY_HOST", "https://software-center.ubuntu.com")

# buy-something base url
#BUY_SOMETHING_HOST = "http://localhost:8000/"
BUY_SOMETHING_HOST = os.environ.get(
    "SOFTWARE_CENTER_AGENT_HOST", SOFTWARE_CENTER_BUY_HOST)

BUY_SOMETHING_HOST_ANONYMOUS = BUY_SOMETHING_HOST

# recommender
RECOMMENDER_HOST = os.environ.get(
    "SOFTWARE_CENTER_RECOMMENDER_HOST", "https://rec.ubuntu.com")

# for the sso login.  ussoc expects the USSOC_SERVICE_URL environment variable
# to be a full path to the service root (including /api/1.0), not just the
# hostname, so we use the same convention for UBUNTU_SSO_SERVICE:
UBUNTU_SSO_SERVICE = os.environ.get(
    "USSOC_SERVICE_URL", "https://login.ubuntu.com/api/1.0")

# the terms-of-service links (the first is for display in a web browser
# as it has the header and footer, the second is for display in a dialog
# as it lacks them and so looks better)
SOFTWARE_CENTER_TOS_LINK = "https://apps.ubuntu.com/cat/tos/"
SOFTWARE_CENTER_TOS_LINK_NO_HEADER = "https://apps.ubuntu.com/cat/tos/plain/ "

# version of the database, every time something gets added (like
# terms for mime-type) increase this (but keep as a string!)
DB_SCHEMA_VERSION = "7"

# the default limit for a search
DEFAULT_SEARCH_LIMIT = 10000

# the server size "page" for ratings&reviews
REVIEWS_BATCH_PAGE_SIZE = 10


# the various "views" that the app has
class ViewPages:
    AVAILABLE = "view-page-available"
    INSTALLED = "view-page-installed"
    HISTORY = "view-page-history"
    SEPARATOR_1 = "view-page-separator-1"
    PENDING = "view-page-pending"
    CHANNEL = "view-page-channel"

    # items considered "permanent", that is, if a item disappears
    # (e.g. progress) then switch back to the previous on in permanent
    # views (LP:  #431907)
    PERMANENT_VIEWS = (AVAILABLE,
                       INSTALLED,
                       CHANNEL,
                       HISTORY
                       )


# define ID values for the various buttons found in the navigation bar
class NavButtons:
    CATEGORY = "category"
    LIST = "list"
    SUBCAT = "subcat"
    DETAILS = "details"
    SEARCH = "search"
    PURCHASE = "purchase"
    PREV_PURCHASES = "prev-purchases"


# define ID values for the action bar buttons
class ActionButtons:
    INSTALL = "install"
    ADD_TO_LAUNCHER = "add_to_launcher"
    CANCEL_ADD_TO_LAUNCHER = "cancel_add_to_launcher"


# icons
class Icons:
    APP_ICON_SIZE = 48

    FALLBACK = "applications-other"
    MISSING_APP = FALLBACK
    MISSING_PKG = "dialog-question"   # XXX: Not used?
    GENERIC_MISSING = "gtk-missing-image"
    INSTALLED_OVERLAY = "software-center-installed"


# sorting
class SortMethods:
    (UNSORTED,
     BY_ALPHABET,
     BY_SEARCH_RANKING,
     BY_CATALOGED_TIME,
     BY_TOP_RATED,
     ) = range(5)


class ReviewSortMethods:
    REVIEW_SORT_METHODS = ['helpful', 'newest']
    REVIEW_SORT_LIST_ENTRIES = [_('Most helpful first'), _('Newest first')]


# values used in the database
class XapianValues:
    APPNAME = 170
    PKGNAME = 171
    ICON = 172
    GETTEXT_DOMAIN = 173
    ARCHIVE_SECTION = 174
    ARCHIVE_ARCH = 175
    POPCON = 176
    SUMMARY = 177
    ARCHIVE_CHANNEL = 178
    DESKTOP_FILE = 179
    PRICE = 180
    ARCHIVE_PPA = 181
    ARCHIVE_DEB_LINE = 182
    ARCHIVE_SIGNING_KEY_ID = 183
    PURCHASED_DATE = 184
    SCREENSHOT_URLS = 185             # multiple urls, comma separated
    ICON_NEEDS_DOWNLOAD = 186         # no longer used
    THUMBNAIL_URL = 187               # no longer used
    SC_DESCRIPTION = 188
    APPNAME_UNTRANSLATED = 189
    ICON_URL = 190
    CATEGORIES = 191
    LICENSE_KEY = 192
    LICENSE_KEY_PATH = 193           # no longer used
    LICENSE = 194
    VIDEO_URL = 195
    DATE_PUBLISHED = 196
    SUPPORT_SITE_URL = 197
    VERSION_INFO = 198
    SC_SUPPORTED_DISTROS = 199
    WEBSITE = 200
    CURRENCY = 201
    # this is used to provide a cataloged time if there is no a-x-i in use
    # or if a-x-i is not available yet
    DB_CATALOGED_TIME = 202
    # download size as the agent provides it
    DOWNLOAD_SIZE = 203


class AppInfoFields:
    ARCH = 'ARCH'
    CHANNEL = 'CHANNEL'
    DEB_LINE = 'DEB_LINE'
    DEB_LINE_ORIG = 'DEB_LINE_ORIG'
    DOWNLOAD_SIZE = 'DOWNLOAD_SIZE'
    CATEGORIES = 'CATEGORIES'
    DATE_PUBLISHED = 'DATE_PUBLISHED'
    DESCRIPTION = 'DESCRIPTION'
    GENERIC_NAME = 'GENERIC_NAME'
    GETTEXT_DOMAIN = 'GETTEXT_DOMAIN'
    ICON = 'ICON'
    ICON_URL = 'ICON_URL'
    IGNORE = 'IGNORE'
    KEYWORDS = 'KEYWORDS'
    LICENSE = 'LICENSE'
    LICENSE_KEY = 'LICENSE_KEY'
    LICENSE_KEY_PATH = 'LICENSE_KEY_PATH'
    MIMETYPE = 'MIMETYPE'
    NAME = 'NAME'
    NAME_UNTRANSLATED = 'NAME_UNTRANSLATED'
    PACKAGE = 'PACKAGE'
    POPCON = 'POPCON'
    PPA = 'PPA'
    PRICE = 'PRICE'
    PURCHASED_DATE = 'PURCHASED_DATE'
    SECTION = 'SECTION'
    SIGNING_KEY_ID = 'SIGNING_KEY_ID'
    SCREENSHOT_URLS = 'SCREENSHOT_URLS'
    SUMMARY = 'SUMMARY'
    SUPPORT_URL = 'SUPPORT_URL'
    SUPPORTED_DISTROS = 'SUPPORTED_DISTROS'
    TAGS = 'TAGS'
    THUMBNAIL_URL = 'THUMBNAIL_URL'
    TYPE = 'TYPE'
    VERSION = 'VERSION'
    VIDEO_URL = 'VIDEO_URL'
    WEBSITE = 'WEBSITE'
    CURRENCY = 'CURRENCY'


# fake channels
AVAILABLE_FOR_PURCHASE_MAGIC_CHANNEL_NAME = "available-for-pay"


# custom keys for the new-apps repository, correspond
# control file custom fields:
#  XB-AppName, XB-Icon, XB-Screenshot-Url, XB-Thumbnail-Url, XB-Category
class CustomKeys:
    APPNAME = "AppName"
    ICON = "Icon"
    SCREENSHOT_URLS = "Screenshot-Url"
    THUMBNAIL_URL = "Thumbnail-Url"
    CATEGORY = "Category"


# pkg action state constants
class PkgStates:
    # current
    INSTALLED = "installed"
    UNINSTALLED = "uninstalled"
    UPGRADABLE = "upgradable"
    REINSTALLABLE = "reinstallable"
    # progress
    INSTALLING = "installing"
    REMOVING = "removing"
    UPGRADING = "upgrading"
    ENABLING_SOURCE = "enabling_source"
    INSTALLING_PURCHASED = "installing_purchase"
    # special
    NEEDS_SOURCE = "needs_source"
    NEEDS_PURCHASE = "needs_purchase"
    PURCHASED_BUT_REPO_MUST_BE_ENABLED = "purchased_but_repo_must_be_enabled"
    ERROR = "error"
    FORCE_VERSION = "force_version"
    # the package is not found in the DB or cache
    NOT_FOUND = "not_found"
    # its purchased but not found for the current series
    PURCHASED_BUT_NOT_AVAILABLE_FOR_SERIES = \
        "purchased_but_not_available_for_series"
    # something went wrong and we don't have a state for this PKG
    UNKNOWN = "unknown"


# visibility of non-applications in the search results
class NonAppVisibility:
    ALWAYS_VISIBLE = "non-apps-always-visible"
    MAYBE_VISIBLE = "non-apps-maybe-visible"
    NEVER_VISIBLE = "non-apps-never-visible"


# application actions
class AppActions:
    INSTALL = "install"
    REMOVE = "remove"
    UPGRADE = "upgrade"
    APPLY = "apply_changes"
    PURCHASE = "purchase"


# transaction types
class TransactionTypes:
    INSTALL = "install"
    REMOVE = "remove"
    UPGRADE = "upgrade"
    APPLY = "apply_changes"
    REPAIR = "repair_dependencies"


# Search separators
class SearchSeparators:
    REGULAR = " "
    PACKAGE = ","


# action values for recommendations feedback
class RecommenderFeedbackActions:
    VIEWED = "viewed"
    INSTALLED = "installed"

# mouse event codes for back/forward buttons
# TODO: consider whether we ought to get these values from gconf so that we
#       can be sure to use the corresponding values used by Nautilus:
#           /apps/nautilus/preferences/mouse_forward_button
#           /apps/nautilus/preferences/mouse_back_button
MOUSE_EVENT_FORWARD_BUTTON = 9
MOUSE_EVENT_BACK_BUTTON = 8

# delimiter for directory path separator in app-install
APP_INSTALL_PATH_DELIMITER = "__"

#carousel app limit to override limit in .menu file for category
TOP_RATED_CAROUSEL_LIMIT = 12
WHATS_NEW_CAROUSEL_LIMIT = 9
LOBBY_RECOMMENDATIONS_CAROUSEL_LIMIT = 9
DETAILS_RECOMMENDATIONS_CAROUSEL_LIMIT = 4

# Transaction ID for the "fake" purchase transaction
PURCHASE_TRANSACTION_ID = "FakePurchaseTransactionID"

from .version import VERSION, DISTRO, RELEASE, CODENAME
WEBKIT_USER_AGENT_SUFFIX = "SoftwareCenter/%s %s/%s (%s)" % (
    VERSION, DISTRO, RELEASE, CODENAME)

# global backend switch, prefer aptdaemon, if that can not be found, use PK
USE_PACKAGEKIT_BACKEND = False
try:
    import aptdaemon
    aptdaemon  # pyflakes
    USE_PACKAGEKIT_BACKEND = False
except ImportError:
    try:
        from gi.repository import PackageKitGlib
        PackageKitGlib  # pyflakes
        USE_PACKAGEKIT_BACKEND = True
    except ImportError:
        raise Exception("Need either aptdaemon or PackageKitGlib")
# allow force via env (useful for testing)
if "SOFTWARE_CENTER_FORCE_PACKAGEKIT" in os.environ:
    USE_PACKAGEKIT_BACKEND = True
